### ISSUE TITLE
Add tags for nodes

### ISSUE THREAD

**Issue by XavM** [hashicorp/consul#154]
Tags can be used with services but not with nodes

Adding tags for nodes and the associated filter for /v1/catalog/nodes HTTP API would be useful

**Comment by jhmartin** [comment]
For fact-retrieval, one might consider integrating with Chef Ohai -- that'd bring in a lot of node facts very easily.

**Comment by johannegger** [comment]
i would also like to express my interest in this feature.

**Comment by dellis23** [comment]
To throw in another use-case along with my :+1:, I could see this being used for tagging things as being in maintenance mode, which would then be provided to consul template for updating configuration.

**Comment by ryanuber** [comment]
@dellis23 Consul 0.5 has built-in support for node-level and service-level maintenance modes. When maintenance mode is enabled, the node or service will be automatically excluded in all queries for nodes/services with passing health checks. Check out the [`maint` command](https://consul.io/docs/commands/maint.html) and [http endpoints](https://consul.io/docs/agent/http/agent.html) to get a feel for how it works!

**Comment by ryanuber** [comment]
I agree with @armon, having node-level tags seems reasonably useful. I think this would provide enough functionality to add "facts" about nodes by use of tags without needing to support specific fact generation systems like Ohai, facter, etc.

**Comment by sitano** [comment]
+1 - we use the same look up patterns based on tags

What do you think if it will be implemented by someone else? (or you would like to impl it yourself)

Actually, i think we can consider any node have a tag if its service have a tag. In that sense, there is also another issue corresponding to nodes look up - there is no easy way to ask consul for a list of nodes that have subset of tags on services they belongs to. What i would like to have too.

Additionally, its not very good i have to ask each DataCenter the same query just to obtain any info across the whole world. It will be good to have something like ?dc=*.

Another thing, there is no way to ask for a nodes list in the DC with full info. Now /catalog/nodes returns almost nothing but simple list. Imagine 100 nodes, how long should i ask consul for tag info on them. And /health/service/{service} which returns nodes, do not have a simple variant of just /health/service request which should be the same as /health/service/*

**Comment by ryanuber** [comment]
Hi @sitano,

> What do you think if it will be implemented by someone else? (or you would like to impl it yourself)

We are always open to implementations from the community! We should agree on the design and functionality first, though, and I don't think we're quite there yet.

> Actually, i think we can consider any node have a tag if its service have a tag.

I think we will run into some odd collision cases by going that route. This would also still require a service to be registered before any tag system would be available, which would promote registering a service just to add tags to the node. While it would work, the UX doesn't quite feel right to me.

> its not very good i have to ask each DataCenter the same query just to obtain any info across the whole world. It will be good to have something like ?dc=*.

We added some plumbing not long ago that would make this type of query possible internally, but honestly from an API perspective I think that the current design works well and lends itself easily to obtaining global information across datacenters. If you have a more specific use case, please open a separate issue and we can discuss it there.

> there is no way to ask for a nodes list in the DC with full info.

I *think* the API's were designed this way to avoid absolutely massive responses. Consider a case where there are hundreds of nodes in the same datacenter. Returning each node with its full list of services and tags inline will be a very large response, and contain lots of duplicated information. As they are, the API's are composable so that they can help clients build any kind of response they need. 

Since the original issue here is about adding tags support to nodes, please open separate issues if you would like to discuss any of the non-related stuff above in more detail.

**Comment by sitano** [comment]
The idea was not to emulate tag system for nodes through applying custom services to them. 

But to have an API which can:
1) get a list of nodes filtered by nodes-tags in certain dc.
2) get a list of nodes filtered by service-tags those nodes are part of in certain dc.

Currently for my production i need (2). But the (2) is a subset of this Issue.

a) For (1),(2) we can extend: /v1/catalog/nodes with ?tag= filter, to filter based on ServiceTags first, and then with NodeTags second, when impl support.

b) For (2) /v1/catalog/service/<service> with ?tag= filter, to filter nodes based on ServiceTags

c) It is better if this ?tag will support something like ?tag=First,Second,LastTag

d) For (2) /v1/catalog/service/<service> to be able to support prefix style request of the form  /v1/catalog/service/[service_prefix]*?tag=blah - what means select all nodes filtered by services which are filtered by wildcard or prefix, and then filter them by ServiceTags they have.

What do you think? I can implement it if we agree on api changes before i start.

**Comment by EricMCornelius** [comment]
I would like to chime in and say that I'm currently evaluating consul to replace some of our home-grown service discovery systems.

One of the biggest pain points I'm currently hitting is the inability for flexible tagging and querying of service instances.

In particular, there's no easy way for us to *version* services with a semver-style major.minor(.patch) and return filtered results of healthy nodes which adhere to a given policy.  It would be extremely useful to have the ability to query richer metadata about services (e.g. where metadata.major_ver=1 and metadata.minor_ver>=3).

We can roll our own protocol on top of the existing service tags, but the restriction that they're a flat array of strings is very limiting in its own right.  It leaves us with little alternative other than to dump stringified json directly into the tags array...

Are there any recommendations on how to tackle this aside from waiting for a richer set of metadata + associated filtering/querying apis?

**Comment by EricMCornelius** [comment]
Note that we could obviously hit the k/v store and use some opaque tag or service unique id, but that's an awful lot of extra complexity for what usually amounts to a simple question: which services can fulfill my request?

I'm happy to tackle contributing a more flexible feature like this back into core if that's the best approach.

**Comment by sitano** [comment]
From my point of view, it looks like, Consul API do not have any space to be extended to support flexible filtering / querying of nodes / services / hc information. It seems like it is not even close to use case patterns HashiCorp is trying to match for Consul. So, i have developed web service with SQL REST endpoint to query in-memory database which is caching all the information from all consuls across all datacenters. It similar to something like facebook`s osquery. I hope someday Consul will support more flexible querying api.

**Comment by ybogdanov** [comment]
+1 for `Node Tags`

@sitano can you share it?

**Comment by sitano** [comment]
@ybogdanov actually i am trying to discuss it with my employer. it turned out quite useful thing for us.

**Comment by ybogdanov** [comment]
@sitano as a temporary solution, I'm thinking of something like `tail -f /dev/null` service on every node, to which tags can be assigned. And this will be a stub `node tags` service, by which you can easily query nodes by tags. Will it work, or am | missing something?

**Comment by sitano** [comment]
@ybogdanov , it is possible. it was the way i was thinking about any workaround for it. but the api which do selects for you is very restricted. my advice - check in advance if it fits your future usage patterns. we were filtering nodes based on tags / services over 10 datacenters, and it took a lot of time. So, then i did write my sql cache.

**Comment by mainframe** [comment]
+1 for Node Tags - would be VERY useful feature

**Comment by ssenaria** [comment]
This is one that is holding us back from adopting Consul. Any idea as to when this will be implemented. Any workarounds you guys can share?

**Comment by armon** [comment]
@ssenaria Likely this would not be until a V2 API which would allow us to change various representations and rethink the APIs a bit. People instead either just create a virtual "service" on every node and attach tags to that or create entries in the KV store.

**Comment by pearkes** [comment]
:+1: this would be great!

**Comment by scalp42** [comment]
Our main use is to have a single shared cluster and use node tags to differentiate between environments.

**Comment by sflint** [comment]
+1 to environments

**Comment by GonZo** [comment]
+1 node tags

**Comment by MatthewHartstonge** [comment]
:+1:  node tags.

**Comment by dky** [comment]
+1 need node tags.

**Comment by plorenz** [comment]
Would it be possible to make serf tags manageable from consul? I would like to be able to use node level tags with serf semantics. 

Or would these node level tags have semantics different from serf tags?

**Comment by JoeOrtiz** [comment]
+1 node tags

**Comment by slackpad** [comment]
Hi @plorenz I don't think we'd implement this using Serf tags. Consul makes use of them internally for some operations, but there are limits to the number of tags, etc. that would make this a tricky feature to use if we exposed Serf tags. I think these tags would be Consul-level and managed between the Consul agent and the catalog.

**Comment by bfgoodrich** [comment]
This would definitely be a useful feature

**Comment by ajvb** [comment]
:+1: Would be useful for me as well. Would like to be able to grab all nodes related to a specific environment (production servers for application X) that span many service definitions.

**Comment by scalp42** [comment]
Went with the approach of having every node with a "generic" /dev/null check every 30-60s and adding tags for the environment if someone wondering about this.

**Comment by modiji** [comment]
+1 for node tags

**Comment by yosefy** [comment]
+1 node tags


becuse:
for inventory on AWS ot Goo
say ansible uses ec2 instances tags as groups
also prometheus may use it as inventory

and all those guys have tags for instance

and then on colocation we install consul and it works in different system
you see the "pain"

**Comment by yosefy** [comment]
hi , is  this topic out of question?
or shoud we try serf for that?
please could you make a bit of clear on that :)
thanks

**Comment by 123BLiN** [comment]
+1

**Comment by phyber** [comment]
This was literally one of the first things I tried to do after having Consul up and running on a set of test nodes, only to find that it's not a feature. Seems odd.

**Comment by ssenaria** [comment]
@yosefy We ended up using Serf but now have a need to use Consul. I can't abandon Serf for Consul and I don't want to have to run both!

**Comment by scalp42** [comment]
Again, use "generic" `/dev/null` checks to work around the issue and you'll have node tags until it's actually implemented.

**Comment by gamefundas** [comment]
I didn't quite follow the /dev/null approach. Does this refer to registering the service with a dummy check and tag the service to our needs? In this case I am assuming the service names can be the same.

Particularly interested in the behavior, where lets say 

Service A was tagged with Tag1, Tag2 & Tag3 in Node1
Service A was tagged with Tag 2, Tag3, Tag4 in Node 2

Will querying for Service A and Tag 1 yield Node 1, Service A and Tag 2 return Node 1 & 2?

**Comment by gamefundas** [comment]
I found out that if both node1 and node2 are registered as Service A with different tagging scheme then Consul rolls up the tags from each node's service definition and forms a global set of tags. So essentially there is no way to apply the filters discretely. Also observed that not all events got propagated, events were getting missed out. 

Alternately I tried adding different service id's for each node and that makes it work. Just that there is multiple registrations for the same service. Another thing to note is that the tag names are matched as contains, for example, tag1, tag2, tag3 can all be matched if one queries for "-tag tag" or "-tag a" etc. This is very strange as one has to define tags that are very much exclusive of each other.

Not sure if I am missing something basic here.
